Celeste.Mod.mm/Mod/Entities/EntityCollider.cs

```diff
@@ -1,26 +1,45 @@
 ﻿using Monocle;
 using System;
 using Microsoft.Xna.Framework;
+using Mono.Cecil;
+using MonoMod.Cil;
 
 namespace Celeste.Mod.Entities {
+    /// <summary>
+    /// Base type of <see cref="EntityCollider{T}"/>.
+    /// In case you don't know the exact type.
+    /// </summary>
+    [Tracked(false)]
+    public abstract class EntityCollider : Component {
+        public abstract Type EntityType { get; }
+
+        public virtual Action<Entity> OnEntityAction => throw new NotImplementedException();
+
+        public Collider Collider;
+
+        public EntityCollider(bool active, bool visible) : base(active, visible) {
+        }
+    }
     /// <summary>
     /// Allows for Collision with any type of entity in the game, similar to a PlayerCollider or PufferCollider.
     /// Performs the Action provided on collision. 
     /// </summary>
     /// <typeparam name="T">The specific type of Entity this component should try to collide with</typeparam>
-    [Tracked(false)]
-    public class EntityCollider<T> : Component where T : Entity {
+    public class EntityCollider<T> : EntityCollider where T : Entity {
         /// <summary>
         /// Provides a simple way to know the Entity type of the specific Collider
         /// </summary>
-        public Type EntityType => typeof(T);
+        public override Type EntityType => typeof(T);
 
         /// <summary>
         /// The Action invoked on Collision, with the Entity collided with passed as a parameter
         /// </summary>
-        public Action<T> OnEntityAction;
+        public new Action<T> OnEntityAction;
 
-        public Collider Collider;
+        [MonoMod.NotPatchEntityColliderMakeVirtual]
+        public Action<Entity> get_OnEntityAction() => e => {
+            OnEntityAction((T) e);
+        };
```

Comment:
What's the purpose of having the action under a wrapper action?

Reply:
Unless this is some MonoMod jank that I don't understand

Reply:
i thought we need to be able to trigger the action without knowing T in code (mainly in reflection), so i added `public Action<Entity> OnEntityAction { get; protected set; }`.
but it is a variable, why not replace it with a get-only property?
we can't override it as there's already a field with the same name, i can only override it in this way.

Reply:
How would replacing work? Do you mind making a commit with the change if you think it's better?